packages/core/src/guards.ts

```diff
@@ -108,14 +108,28 @@ export function toGuardDefinition<
   TEvent extends EventObject
 >(
   guardConfig: GuardConfig<TContext, TEvent>,
-  getPredicate?: (guardType: string) => GuardPredicate<TContext, TEvent>
+  getPredicate?: (
+    guardType: string
+  ) => GuardPredicate<TContext, TEvent> | GuardDefinition<TContext, TEvent>
 ): GuardDefinition<TContext, TEvent> {
+  // TODO: do this recursively and check for cycles
   if (isString(guardConfig)) {
```

Comment:
All of this feels a tad complicated - but it's not exactly a unique problem to this function (similar things happen when "resolving" actions).

I think that the main mistake that we do here is that we try to resolve (through `toGuardDefinition`) too eagerly. We always need to have this resolution~ step when finally checking guards since we need to be inside the execution context of a machine. It's kinda pointless to resolve this twice.

Let's roll with this for now but I think that this should be flagged as something to refactor.